### ISSUE TITLE
Bayesian principal-stratification IV for non-compliance / LATE estimation

### ISSUE THREAD

**Issue by PabloRoque** [pymc-labs/CausalPy#1032]
## Summary

CausalPy has strong Bayesian IV support, but it models endogeneity as a **continuous latent confounder** and cannot represent **non-compliance** as discrete principal strata or target a **LATE** (complier-average causal effect). This is a feature request to add a principal-stratification IV model for the binary-instrument / binary-treatment-received / count-outcome case under monotonicity.

## Current functionality

Bayesian IV is provided via `InstrumentalVariable` (`causalpy/experiments/instrumental_variable.py`) + `InstrumentalVariableRegression` (`causalpy/pymc_models.py`):

- Joint simultaneous-equations model (bivariate-normal / `LKJCholeskyCov`) estimating a structural effect while modeling treatment–outcome residual correlation as the endogeneity channel.
- A `binary_treatment=True` control-function branch (logit treatment stage + `rho` correlation parameter).
- Variable-selection priors (spike-and-slab / horseshoe) for high-dimensional instrument/covariate sets.
- Notebooks for basic IV, weak instruments, and variable-selection priors.

## The gap

The existing IV model treats endogeneity as a continuous correlated residual. It does **not** model non-compliance as discrete strata, and cannot target a LATE. Concretely, it's missing:

1. **Discrete compliance types.** No representation of never-taker / complier / always-taker strata under monotonicity, and therefore no separation of the instrument's effect on *take-up* from its effect on the *outcome*. The current model conflates "instrument barely moves treatment" with "treatment barely moves outcome."
2. **LATE as an estimand.** The output is a single structural coefficient (an ATE-like quantity under its assumptions), not a complier-average effect. There is no posterior over the complier share and no per-unit / per-covariate-profile `P(complier | data)`. "LATE" and "complier" appear only as prose in `docs/source/knowledgebase/estimands.md` and the weak-instruments notebook — nothing implements them.
3. **Count (and other non-Gaussian) outcomes.** Both IV branches assume a Normal outcome. There is no Poisson/NegativeBinomial path, so a natural outcome like *number of orders per customer* can only be handled as a misspecified continuous approximation.
4. **No discrete-latent-variable machinery at all.** There is currently zero use of `pm.Mixture`, `logsumexp`, or any marginalized-latent-class pattern anywhere in the codebase — a mixture-over-strata likelihood is genuinely new ground, not a variant of an existing model.

## Why it matters

The motivating case is estimating promotion / marketing-channel effects on customer order counts, where:

- **Non-compliance is the norm** — an encouragement/eligibility instrument shifts *who takes up* the promo, it does not assign treatment deterministically. The estimand of interest is the effect on customers who actually respond (compliers), plus the size of each stratum (e.g. always-takers we're spending promo budget on for nothing).
- **Instruments are typically weak** — and the correct signature of a weak instrument here is a complier share `pi_C` concentrated near zero, which the current continuous-residual model has no way to surface.

## Scope

Initial scope: the **binary-instrument / binary-treatment-received / count-outcome** case under **monotonicity** (no defiers → three strata). Continuous-dose analogs, hierarchical multi-market pooling, full type × covariate outcome interactions, and relaxing monotonicity (defiers) are natural follow-ups, not part of the first cut.

A proposed solution / design sketch follows in a comment below, so the API and modeling choices can be discussed in-thread.

**Comment by PabloRoque** [comment]
## Proposed solution

Two new pieces, following existing CausalPy conventions where sensible and deviating only where the discrete-latent-class marginalization requires it.

### Model — `PrincipalStratificationRegression(PyMCModel)`

New class in `causalpy/pymc_models.py`, after `InstrumentalVariableRegression`.

```python
def build_model(self, X, W, Z, D, y, coords, priors,
                outcome_dist="poisson",           # or "negativebinomial"
                vs_prior_type=None, vs_hyperparams=None): ...
```

- `X`: outcome-rate covariates. `W`: type-membership covariates. `Z`: binary instrument. `D`: binary treatment received. `y`: count outcome.
- **Type-membership submodel:** softmax regression on `W`, with `always-taker` pinned as the reference category (logit = 0), so only `beta_NT` and `beta_C` are free. This fixes label identity **by construction** — the three types are non-exchangeable, so no post-hoc relabeling/sorting is needed (or wanted).
- **Outcome submodel:** log-link count GLM (`Poisson` or `NegativeBinomial`) with a shared `X @ beta_y0`, plus type-specific intercept shifts `gamma_NT`, `gamma_AT`, and `tau_LATE` as the complier D=0→D=1 effect. Intercept-only per type for v1 (not full type × covariate interaction) to bound the parameter count in an already weak-instrument-prone setting.
- **Marginalized likelihood.** Under monotonicity, each observed `(Z, D)` cell maps to either a single deterministic type or a two-component mixture:

  | Cell | Consistent types |
  |------|------------------|
  | `Z=0, D=0` | {never-taker, complier@D=0} |
  | `Z=0, D=1` | always-taker (deterministic) |
  | `Z=1, D=0` | never-taker (deterministic) |
  | `Z=1, D=1` | {complier@D=1, always-taker} |

  Per-unit log-likelihood (note the deterministic cells still carry the stratum-probability factor `pi_AT` / `pi_NT`, which is what actually identifies the shares from compliance behavior):

  ```
  log L_i =
    1[Z=0,D=0] * logsumexp(log(pi_NT) + logp_NT, log(pi_C) + logp_C0)
  + 1[Z=0,D=1] * (log(pi_AT) + logp_AT)
  + 1[Z=1,D=0] * (log(pi_NT) + logp_NT)
  + 1[Z=1,D=1] * logsumexp(log(pi_C) + logp_C1, log(pi_AT) + logp_AT)
  ```

  Implemented hand-rolled via `pm.math.logsumexp` over `pt.stack([...])`, masked per cell with `pt.switch` on `Z`/`D` (not boolean fancy-indexing — keeps the graph shape-stable for `pm.Data`), added with `pm.Potential("likelihood", logl.sum())`.

- **Deliberately not `pymc_extras.MarginalModel`** (even though `pymc-extras` is already a core dep): its fixed transition/emission structure doesn't fit this data-dependent, per-cell masked mixture, and every existing custom likelihood in `pymc_models.py` is hand-rolled. Hand-rolling keeps the diff small and auditable — which matters because debugging weak-instrument identifiability will be most of the work.
- Exposes `pm.Deterministic("pi", ...)` (stacked type shares) and `pm.Deterministic("LATE", lambda_C1 - lambda_C0)`. `P(complier | data)` per unit/profile is computed **post-hoc** via Bayes' rule on the stored components (a `posterior_complier_probability(...)` method), keeping `build_model` a training-time-only graph.
- `NegativeBinomial` adds a shared dispersion `alpha`. Variable-selection priors reuse `VariableSelectionPrior` exactly as `InstrumentalVariableRegression` does, applied independently to `beta_NT`/`beta_C` and `beta_y0`.
- No out-of-sample `predict()` in v1 (follow-up).

### Experiment — `NonComplianceInstrumentalVariable(BaseExperiment)`

New file `causalpy/experiments/noncompliance_iv.py`; `supports_ols = False`, `supports_bayes = True`.

```python
def __init__(self, instruments_data, data,
             instruments_formula,   # e.g. "Z ~ 1"
             treatment_formula,     # e.g. "D ~ 1 + w1 + w2"  -> W (type membership)
             formula,               # e.g. "y ~ 1 + x1"       -> X (outcome rate)
             model=None, priors=None, outcome_dist="poisson",
             vs_prior_type=None, vs_hyperparams=None, **kwargs): ...
```

- Three patsy formulas via the existing `build_formula_matrices` (`causalpy/formula_utils.py`).
- `input_validation()`: `DataException` on non-binary `Z`/`D` (mirrors the existing binary check in `instrumental_variable.py`); a soft `warnings.warn` if the raw cross-tab shows `P(D=1|Z=1) < P(D=1|Z=0)` — the empirical monotonicity-violation signature.
- `algorithm()`: cheap classical fits (compliance cross-tabs + naive per-cell Poisson GLM) to set informative-but-not-flat prior defaults — same "classical fit informs priors" convention as `InstrumentalVariableRegression.algorithm()`.
- **`plot()` and `effect_summary()` are implemented for real** (unlike `InstrumentalVariable`, where both are stubs): `effect_summary()` reports the LATE (mean/HDI/tail prob) and `pi_C` (flagging weak-instrument risk when its HDI sits near 0); `plot()` shows the `pi` posterior, the LATE posterior, and the per-profile `P(complier)` distribution.
- **Diagnostics are first-class:** `summary()` surfaces `rhat`/`ess_bulk` for `beta_NT`, `beta_C`, `tau_LATE`, `gamma_AT` with an explicit warning on threshold breach — the operational symptom of weak-instrument near-non-identification in this parameterization.

### Tests — `causalpy/tests/test_noncompliance_iv.py`

- Two synthetic DGP fixtures (via the existing `rng` fixture): one well-identified (known strata shares + `tau_LATE`), one deliberately weak-instrument (shares barely differ by `Z`).
- Init/shape tests; `input_validation` (non-binary → `DataException`, monotonicity violation → `pytest.warns`); parameter recovery (HDI contains truth for `pi_C` and `LATE`, full recovery behind the suite's slow marker).
- Weak-instrument test: fit runs without NaNs; the diagnostic-warning threshold logic is unit-tested as a pure function against a synthetic low-ESS `idata` (rather than relying on a real MCMC run to reliably produce low ESS in CI).
- Positive tests for `effect_summary()`/`plot()` and `posterior_complier_probability()`.

### Docs

- New notebook `docs/source/notebooks/instrumental-variables-noncompliance-late.ipynb` (promo → order-count → weak-instrument walkthrough), registered in `docs/source/notebooks/gallery.yaml` under the existing "Instrumental Variables Regression" section (`index.md` is generated by `scripts/generate_gallery.py`, not hand-edited).

### Out of scope (follow-ups)

Continuous-dose analog; hierarchical multi-market pooling of `pi`/`tau_LATE`; full type × covariate outcome interactions; relaxing monotonicity (4-type / defiers); out-of-sample prediction.

**Comment by NathanielF** [comment]
Love this idea @PabloRoque. Would keen to see it. Let me know if you want PR or just want to brainstorm

**Comment by drbenvincent** [comment]
causalpy-triage-agent: Added labels: `enhancement`. This issue is a feature request to add a principal-stratification IV model.
